### PR TITLE
enableCrushUpdates option: rook operator would not be able to update crush rules by default if not set

### PR DESCRIPTION
**Pull request by ideepika**
Resolves https://github.com/rook/rook/issues/13854 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by BlaineEXE** [comment]
Resolves #13854 ?

**Comment by travisn** [comment]
> A comment from huddle for us to consider:
> 
> We should consider whether `disableCrushUpdates` or `enableCrushUpdates` is more appropriate. Choosing `enable...` would mean that we are changing existing behavior (might confuse/frustrate existing users of this feature), but we would also make the default configuration one that guarantees a more stable Ceph cluster (would be less likely to result in surprise data movement).
> 
> I think we could choose to make either decision, but let's make sure we are clear about which scenarios we are prioritizing.

Unexpected updating of crush rules has been a point of concern for several users particularly since #13069 was merged. Therefore, I would also suggest we disable updates by default, and name the setting `enableCrushUpdates`.

**Comment by travisn** [comment]
You may need to run `make docs` to fix the CI

**Comment by ideepika** [comment]
working as expected: `2024-06-12 16:18:47.758207 D | cephclient: Skipping crush rule update for pool "replicapool": EnableCrushUpdates is disabled `
```
        {
            "rule_id": 1,
            "rule_name": "replicapool",
            "type": 1,
            "steps": [
                {
                    "op": "take",
                    "item": -1,
                    "item_name": "default"
                },
                {
                    "op": "choose_firstn",
                    "num": 0,
                    "type": "osd"
                },
                {
                    "op": "emit"
                }
            ]
        },
```

when running with enableCrushUpdates: 
```
      {
            "rule_id": 3,
            "rule_name": "pooltestenablecrush",
            "type": 1,
            "steps": [
                {
                    "op": "take",
                    "item": -2,
                    "item_name": "default~hdd"
                },
                {
                    "op": "chooseleaf_firstn",
                    "num": 0,
                    "type": "host"
                },
                {
                    "op": "emit"
                }
            ]
        },
        {
            "rule_id": 4,
            "rule_name": "pooltestenablecrush_host_hdd",
            "type": 1,
            "steps": [
                {
                    "op": "take",
                    "item": -2,
                    "item_name": "default~hdd"
                },
                {
                    "op": "chooseleaf_firstn",
                    "num": 0,
                    "type": "host"
                },
                {
                    "op": "emit"
                }
            ]
        }
    ]
```
```
2024-06-12 16:42:15.095679 I | cephclient: updating pool "pooltestenablecrush" failure domain from "host" to "host" with new crush rule "pooltestenablecrush_host_hdd"
  2352 2024-06-12 16:42:15.095727 I | cephclient: crush rule "pooltestenablecrush" will no longer be used by pool "pooltestenablecrush"
```

File: deploy/examples/pool.yaml [hunk 1]
```diff
@@ -33,6 +33,8 @@ spec:
   # If device classes are specified, ensure this property is added to every pool in the cluster,
   # otherwise Ceph will warn about pools with overlapping roots.
   #deviceClass: my-class
+  # allow rook to update pool spec and crush rule, use with caution, disabled by default
+  #enableCrushUpdates: false
```
[the remark this answers is not in the document]
Comment (ideepika) [reply deploy/examples/pool.yaml]: done, thanks!

File: pkg/daemon/ceph/client/pool.go [hunk 2]
```diff
@@ -463,6 +463,12 @@ func createReplicatedPoolForApp(context *clusterd.Context, clusterInfo *ClusterI
 }
 
 func updatePoolCrushRule(context *clusterd.Context, clusterInfo *ClusterInfo, clusterSpec *cephv1.ClusterSpec, pool cephv1.NamedPoolSpec) error {
+
+	if !pool.EnableCrushUpdates {
+		logger.Debugf("Skipping crush rule update for pool %q: EnableCrushUpdates is disabled", pool.Name)
```
[the remark this answers is not in the document]
Comment (travisn) [reply pkg/daemon/ceph/client/pool.go:468]: If it is changed to info, I'd suggest a different message such as:
```suggestion
		logger.Infof("Skipping check to see if crush rule update needed for pool %q. EnableCrushUpdates is disabled", pool.Name)
```

[the remark this answers is not in the document]
Comment (BlaineEXE) [reply pkg/daemon/ceph/client/pool.go:468]: I see. In context, this is just a note about checking, not about changing itself. Given that, I'll merge this as-is.

**Comment by BlaineEXE** [comment]
Because this is changing default behavior, let's assume this will just be changed for 4.15 for now.

File: pkg/daemon/ceph/client/pool.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ideepika) [reply pkg/daemon/ceph/client/pool.go:468]: sorry, checked a bit late, thanks for the merge!

**Pull request merged**